src/wfs/src/main/java/org/geoserver/wfs/json/ComplexGeoJsonWriter.java

```diff
@@ -0,0 +1,377 @@
+/* (c) 2017 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.wfs.json;
+
+import com.vividsolutions.jts.geom.Geometry;
+import org.geotools.feature.FeatureCollection;
+import org.geotools.feature.FeatureIterator;
+import org.geotools.referencing.CRS;
+import org.opengis.feature.Attribute;
+import org.opengis.feature.ComplexAttribute;
+import org.opengis.feature.Feature;
+import org.opengis.feature.Property;
+import org.opengis.feature.type.ComplexType;
+import org.opengis.feature.type.GeometryDescriptor;
+import org.opengis.feature.type.Name;
+import org.opengis.feature.type.PropertyDescriptor;
+import org.opengis.feature.type.PropertyType;
+import org.opengis.filter.identity.Identifier;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * GeoJSON writer capable of handling complex features.
+ */
+final class ComplexGeoJsonWriter {
```

Comment:
Yes, there is nothing wrong with a package-private class, and it can be a good idea to limit the visibility of code that could be considered new API, but implementation helper classes are often implemented as nested classes in GeoTools/GeoServer. In my view, if a class merits a separate file, it might be useful enough to be public.  :-)